_posts/pl/2024-04-19-czy-wiesz-jak-uzywac-operatorow-w-javascripcie-i-typescripcie.md

```diff
@@ -0,0 +1,90 @@
+```typescript
+let dlugosc: number = zmienna.length;  // Błąd kompilacji, ponieważ zmienna może być null
+let dlugosc: number = zmienna!.length; // Kompilator taką linijkę kodu przepuści
+```
+Jest to do pewnego stopnia obejście mechanizmu sprawdzania typów i powinno się go używać z rozwagą. Kod taki może być mniej czytelny, a kompilator przestaje kontrolować miejsca, które mogą spowodować wystąpienie błędu w czasie działania programu. Rozważniejszym podejściem jest weryfikowanie, czy zmienna nie ma wartości `null` lub `undefined` i osobno obsługiwanie takich sytuacji.
+
+Dodatkowo, już w czystym Javascript, postawienie `!` przed zmienną sprawia, że działa on jako operator negacji logicznej. Kiedy używamy go przed wyrażeniem lub zmienną, neguje on wartość logiczną tego wyrażenia. Na przykład:
+```javascript
+let prawda = true;
+let falsz = !prawda; // falsz = false
+```
+
+Operator neguje wartość naszej zmiennej, dodatkowo konwertując ją na zmienną typu boolean. Wiedząc to, możemy używać podwójnego zaprzeczenia `!!` jako sposób konwersji wartości na korespondującą z nią wartość boolean. Poniżej kilka przykładów:
+```javascript
+!!""        // zwraca false
+!!"wartość" // zwraca true
+!!0         // zwraca false
+!!1         // zwraca true
+!![1, 2]    // zwraca true
+!![]        // zwraca... true
+```
+W kodzie rzadko raczej trafimy na wystąpienie `!!`, ale warto wiedzieć, co to oznacza.
+
+Pozostał nam jeszcze znak zapytania. Zacznijmy od Typescript. Kiedy definiujemy interfejs możemy posłużyć się nim, aby wskazać, które pola są opcjonalne:
+```typescript
+interface Person {
+  name: string;
+  age?: number; // age jest opcjonalne
+}
+ 
+const person: Person = { name: 'John' }; // Nie powoduje błędu kompilacji
+```
+
+Z kolei w JavaScript możemy posłużyć nam jako operator warunkowy do zapisywania wyrażenia `if-else` w skróconej formie:
+```javascript
+const age = 20;
+const canVote = age >= 18 ? true : false;
+```
+
+W przypadku prostych warunków, które weryfikują istnienie jakiejś wartości warto znać operator `??` (null coalescing operator), który zwróci wartość po lewej stronie operatora jeśli jest ona różna od `null` i `undefined`, a w przeciwnym wypadku zwróci wartość po prawej stronie.
+```javascript
+null ?? "hello"      // "hello"
+undefined ?? "hello" // "hello"
+"world" ?? "hello"   // "world"
+"" ?? "hello"        // ""
+0 ?? "hello"         // 0
+```
+Warto tutaj wspomnieć, że operator alternatywy `||` działa bardzo podobnie, ale dla dwóch ostatnich przykładów zwróciłby inną wartość (opiera się on na wartości logicznej a nie `null` czy `undefined`), więc nie można tych operatorów stosować zamiennie.
+```javascript
+"" || "hello"        // "hello"
+0 || "hello"         // "hello"
+```
+
+Znak zapytania jest używany również jako operator zabezpieczający przed błędami, który pozwala na dostęp do właściwości obiektu, nawet jeśli nie jest ona zdefiniowana (optional chaining).
```

Comment:
tu mam wątpliwość, czy "właściwość" obiektu jest  jedna, czy to "właściwości" w l. mn.
wówczas -> "...nawet jeśli nie są one zdefiniowane"

Reply:
liczba pojedyncza, jest OK